### PR TITLE
Nodejs - Initial nodejs server

### PR DESCRIPTION
**Pull request by kcurtis25**
This is the start of a nodejs server. I am using a technology stack of 'express' (http://expressjs.com/) and jade (http://jade-lang.com/). 

Much of this I did not write from scratch. I generated with the express generator (http://expressjs.com/en/starter/generator.html). 

Once you have nodejs installed, I start it with `$ npm start`, which the generator page above says is the same command for Windows (advantage of nodejs over perl). 

Then navigate a browser to: http://localhost:3000/ and you should see the index page. 

Aside from seeing the basic routes/views split, the only really interesting thing is that `routes/generate.js` loads up an instance of `number-puzzles/index.js` and calls `getPuzzleTypes()` on it to get an array of the names of the puzzle types.

Lots to do on the perl front before I can move much further forward on this.

### CODE REVIEW

File: puzzle-app/app.js [hunk 1]
```diff
@@ -0,0 +1,63 @@
+var express       = require('express');
+var path          = require('path');
+var favicon       = require('serve-favicon');
+var logger        = require('morgan');
+var cookieParser  = require('cookie-parser');
+var bodyParser    = require('body-parser');
+
+var routes        = require('./routes/index');
```
Comment (kcurtis25) [inline puzzle-app/app.js]: So, pretty much, for each URL we want to support, we need to add a `var urlname = require('./routes/urlfilename');` line here. The variables are used on lines 27-30. Is is lines 27-30 where the routes file is linked with the actual URL.

File: puzzle-app/app.js [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+app.use(logger('dev'));
+app.use(bodyParser.json());
+app.use(bodyParser.urlencoded({ extended: false }));
+app.use(cookieParser());
+app.use(express.static(path.join(__dirname, 'public')));
+
+app.use('/', routes);
+app.use('/users', users);
+app.use('/about', about);
+app.use('/generate', generate);
```
Comment (kcurtis25) [inline puzzle-app/app.js]: So, here we could add `app.use('/puzzle-generator', generate);` and that means that when people go to the `/puzzle-generator` URL it uses the same code path as the `/generate` URL. This allows a separation or mapping between what the users see and what code is run.

For sanity, I have kept the names consistent throughout so we don't need to remember the mappings.

File: puzzle-app/number-puzzles/index.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (kcurtis25) [inline puzzle-app/number-puzzles/index.js]: This is going to be the interesting file. Once I have done a reasonable amount of work on perl, this function will look more like below with a http call to the perl (dancer/PSGI) api.

```
var http = require('http');
http.get('http://localhost:8123/number-puzzle/puzzle_type/', (...)
return types;
```

File: puzzle-app/package.json [hunk 4]
```diff
@@ -0,0 +1,17 @@
+{
```
Comment (kcurtis25) [inline puzzle-app/package.json]: This file lists all the packages you need. It is used by `npm`, so you can type `npm install` and npm knows what you need and what version. It is updated when you say `npm install <package_name> --save`. So if you end up adding any packages, make sure you use `--save` so this is updated too. 

Packages are stored locally in your `node_modules` directory. You do not need to be root to install the packages. There is a `.gitignore` file which ensures this directory is ignored by git.

File: puzzle-app/routes/generate.js [hunk 5]
```diff
@@ -0,0 +1,17 @@
+var express   = require('express');
```
Comment (kcurtis25) [inline puzzle-app/routes/generate.js]: Here I am using the same url (`/generate`) to handle both the get (which displays the form) and the post (which will one day process the data). Not sure if we'll stick with that, but I wanted to try it out and so far it works. 

No actual data processing gets done, but the form/request parameters are printed to the console log. Longer term the `post` function will use `puzzle` and call something like 

```
var filename = puzzle.generatePuzzle(req.body);
```

And then display the filename to the user for download (from the static location).

File: puzzle-app/routes/users.js [hunk 6]
```diff
@@ -0,0 +1,9 @@
+var express   = require('express');
```
Comment (kcurtis25) [inline puzzle-app/routes/users.js]: This file came as part of the generated data. We can delete it soon, as I don't think it serves a purpose for us. I think it is an example of how you can return data without a file in the `views` directory

**Comment by phowlin** [comment]
Sets up a good structure of node.js in which to work. Possibly more complicated than a project of this size requires, but that's part of the point to learn good practice even at this level. Thanks for putting this together kcurtis25.

**Pull request merged**